packages/fluentui/react-northstar/src/components/Tree/hooks/useTreeSelectState.tsx

```diff
@@ -0,0 +1,136 @@
+    const updatedFlatTree = {};
+    const rootsIds = [];
+    Object.keys(flatTree).forEach(id => {
+      updatedFlatTree[id] = {
+        ...flatTree[id],
+        selected: false,
+      };
+      if (!updatedFlatTree[id].parent) {
+        rootsIds.push(id);
+      }
+    });
+    selectedItemIds.forEach(id => {
+      if (!updatedFlatTree[id].hasSubtree) {
+        updatedFlatTree[id].selected = true;
+      }
+    });
+
+    // traverse all tree nodes in updatedFlatTree for once,
+    // to calculate the selection state of the parent nodes based on leaf nodes
+    const traverseTree = nodes => {
+      let allNodesSelected = true;
+      let noNodeSelected = true;
+
+      nodes.forEach(id => {
+        if (updatedFlatTree[id].childrenIds) {
+          updatedFlatTree[id].selected = traverseTree(updatedFlatTree[id].childrenIds);
+        }
+
+        if (updatedFlatTree[id].selected === true) {
+          noNodeSelected = false;
+        } else if (updatedFlatTree[id].selected === 'indeterminate') {
+          allNodesSelected = false;
+          noNodeSelected = false;
+        } else {
+          allNodesSelected = false;
+        }
+      });
+
+      if (allNodesSelected) return true;
+      if (noNodeSelected) return false;
+      return 'indeterminate';
+    };
+
+    traverseTree(rootsIds);
+    return updatedFlatTree;
+  }, [flatTree, selectedItemIds]);
+
+  const toggleSelectOnOneId = React.useCallback(
+    (ids: string[], id: string): string[] => {
+      const leafs = getLeavesOfSubTree(updatedFlatTree, id);
+
+      if (updatedFlatTree[id]?.selected === true) {
+        // remove all leaves from selected
+        return leafs.reduce((prevResult, leaf) => {
+          const leafIndex = findIndex(prevResult, leaf);
+          if (leafIndex >= 0) {
+            return removeItemAtIndex(prevResult, leafIndex);
+          }
+          return prevResult;
+        }, ids);
+      }
+
+      return leafs.reduce((prevResult, leaf) => {
+        // add all leaves to selected
+        const leafIndex = findIndex(prevResult, leaf);
+        if (leafIndex < 0) {
+          return [...prevResult, leaf];
+        }
+        return prevResult;
+      }, ids);
+    },
+    [updatedFlatTree],
+  );
+
+  const stableProps = useStableProps(props);
+  const toggleSelect = React.useCallback(
+    (ids: string[], e: React.SyntheticEvent) => {
+      const nextSelectedItemIds = ids.reduce((prev, currId) => toggleSelectOnOneId(prev, currId), selectedItemIds);
+
+      _.invoke(stableProps.current, 'onSelectedItemIdsChange', e, {
+        ...stableProps.current,
+        selectedItemIds: nextSelectedItemIds,
+      });
+
+      setSelectedItemIdsState(nextSelectedItemIds);
+    },
+    [selectedItemIds, setSelectedItemIdsState, stableProps, toggleSelectOnOneId],
+  );
+
+  return { selectedItemIds, flatTree: updatedFlatTree, toggleSelect };
+}
+
+function getLeavesOfSubTree(flatTree, subTreeRootId) {
```

Comment:
nit: the fact that it is a subtree is kind of irrelevant
```suggestion
function getLeafNodes(flatTree, rootId) {
```